crates/fixedpointmath/src/lib.rs

```diff
@@ -1,422 +1,406 @@
 mod macros;
+mod sign;
 mod utils;
+mod value;
 
+use core::panic;
 use std::{
     fmt,
-    ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Shr, Sub, SubAssign},
-    str::FromStr,
+    ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign},
+    os::macos::raw,
 };
 
-use ethers::types::{Sign, I256, U256};
-use eyre::{eyre, ErrReport, Error, Result};
+use ethers::types::{I256, U256, U512};
+use eyre::{bail, eyre, Error, Report, Result};
 use rand::{
     distributions::{
-        uniform::{SampleBorrow, SampleUniform, UniformSampler},
+        uniform::{SampleBorrow, SampleUniform, UniformFloat, UniformSampler},
         Distribution, Standard,
     },
     Rng,
 };
+pub use sign::*;
 pub use utils::*;
+pub use value::*;
 
-/// A fixed point wrapper around the `U256` type from ethers-rs.
+/// A fixed point wrapper around ethers-rs.
+///
+/// This fixed point type is a heavily based on Solidity's FixedPointMath
+/// library with a few changes:
+/// - Support for negative numbers.
+/// - Support for overflowing intermediate operations in `mul_div_down` and
+///  `mul_div_up`.
 ///
-/// This fixed point type is a direct port of Solidity's FixedPointMath library.
 /// Each of the functions is fuzz tested against the Solidity implementation to
-/// ensure that the behavior is identical.
-#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
-pub struct FixedPoint(U256);
-
-impl Default for FixedPoint {
-    fn default() -> FixedPoint {
-        fixed!(0)
-    }
+/// ensure that the behavior is identical given values bounded by both the
+/// Solidity and Rust limits.
+#[derive(PartialOrd, Ord, Clone, Copy)]
+pub struct FixedPoint<T: FixedPointValue> {
+    sign: FixedPointSign,
+    value: T,
+    decimals: u8,
```

Comment:
Why are we storing this with the value? It’s just copying around data that won’t be used unless we display it.

Reply:
Nvm, I see that it’s being used to support fixed point addition with a different decimal point. We could do this before mulDivDown (since it’s just a*b/c), but I see how this could be nicer in some cases